keanu-project/src/main/java/io/improbable/keanu/plating/loop/LoopBuilder.java

```diff
@@ -0,0 +1,180 @@
+package io.improbable.keanu.plating.loop;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+
+import io.improbable.keanu.plating.Plate;
+import io.improbable.keanu.plating.PlateBuilder;
+import io.improbable.keanu.plating.Plates;
+import io.improbable.keanu.vertices.ConstantVertex;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexLabel;
+import io.improbable.keanu.vertices.VertexLabelException;
+import io.improbable.keanu.vertices.bool.BoolVertex;
+import io.improbable.keanu.vertices.bool.nonprobabilistic.BoolProxyVertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.DoubleProxyVertex;
+import io.improbable.keanu.vertices.generic.nonprobabilistic.If;
+
+public class LoopBuilder {
+    private final Collection<Vertex> initialState;
+    private ImmutableMap.Builder<VertexLabel, VertexLabel> customMappings = ImmutableMap.builder();
+    private int maxLoopCount = Loop.DEFAULT_MAX_COUNT;
+    private boolean throwWhenMaxCountIsReached = true;
+
+    <V extends Vertex<?>> LoopBuilder(Collection<V> initialState) {
+        this.initialState = ImmutableList.copyOf(initialState);
+    }
+
+    /**
+     * An optional method to override the default value
+     *
+     * @param maxCount - the max number of times the loop can run
+     * @return self
+     */
+    public LoopBuilder atMost(int maxCount) {
+        this.maxLoopCount = maxCount;
+        return this;
+    }
+
+    /**
+     * An optional method to override the default behaviour
+     * If the maximum loop count is exceeded, it will log a warning instead of throwing
+     *
+     * @return self
+     */
+    public LoopBuilder dontThrowWhenMaxCountIsReached() {
+        this.throwWhenMaxCountIsReached = false;
+        return this;
+    }
+
+    /**
+     * An optional method to add custom mappings
+     *
+     * @param inputLabel  - the label assigned to the ProxyVertex in frame t
+     * @param outputLabel - the label assigned to a Vertex in frame t-1 which will become the ProxyVertex's parent
+     * @return self
+     */
+    public LoopBuilder mapping(VertexLabel inputLabel, VertexLabel outputLabel) {
```

Comment:
I think the confusion is because you're thinking of it as the input/output to the proxy vertex, while I'm thinking about it as the input/output of (two different) frames.

We could rename `inputLabel` as `proxyLabel` and `outputLabel` as `labelFromPreviousFrame` or `proxysParentLabel`?

It would also help to switch the order of the 2 lines of javadoc, e.g. this is what it would look like without any name changes:

     * @param outputLabel the label assigned to a Vertex in frame t-1 which will become the ProxyVertex's parent
     * @param inputLabel  the label assigned to the ProxyVertex in frame t

Reply:
I've done the rename